contracts/upgradeable_contracts/arbitrary_message/BasicForeignAMB.sol

```diff
@@ -0,0 +1,115 @@
+pragma solidity 0.4.24;
+
+import "./BasicAMB.sol";
+import "../../libraries/Message.sol";
+
+
+contract BasicForeignAMB is BasicAMB {
+    uint256 internal constant PASS_MESSAGE_GAS = 100000;
+    address internal accountForAction = address(0);
+
+    event RelayedMessage(address sender, address executor, bytes32 transactionHash);
+
+    function executeSignatures(uint8[] vs, bytes32[] rs, bytes32[] ss, bytes _data) external onlyValidator {
+        Message.hasEnoughValidSignatures(_data, vs, rs, ss, validatorContract());
+
+        processMessage(_data);
+    }
+
+    function withdrawFromDeposit(address _recipient) public {
+        require(msg.sender == address(this));
+        require(accountForAction != address(0));
+        require(balanceOf(accountForAction) > 0);
+        uint256 withdrawValue = balanceOf(accountForAction);
+        setBalanceOf(accountForAction, 0);
+        _recipient.transfer(withdrawValue);
+        accountForAction = address(0);
+    }
+
+    function depositForContractSender(address _contract) public payable {
+        require(_contract != address(0));
+        setBalanceOf(_contract, balanceOf(_contract) + msg.value);
+    }
+
+    function relayedMessages(bytes32 _txHash) public view returns(bool) {
+        return boolStorage[keccak256(abi.encodePacked("relayedMessages", _txHash))];
+    }
+
+    function balanceOf(address _balanceHolder) public view returns(uint) {
+        return uintStorage[keccak256(abi.encodePacked("balances", _balanceHolder))];
+    }
+
+    function isWithdrawFromDepositSelector(bytes _data) internal pure returns(bool _retval) {
+        _retval = false;
+        bytes4 withdrawFromDepositSelector = this.withdrawFromDeposit.selector;
+        if ((_data[0] == withdrawFromDepositSelector[0]) &&
+            (_data[1] == withdrawFromDepositSelector[1]) &&
+            (_data[2] == withdrawFromDepositSelector[2]) &&
+            (_data[3] == withdrawFromDepositSelector[3])) {
+            _retval = true;
+        }
+    }
+
+    function _passMessage(address _sender, address _contract, bytes _data, uint256 _gas) internal {
+        if (_contract == address(this)) {
+            //Special check to handle invocation of withdrawFromDeposit
+            if (isWithdrawFromDepositSelector(_data)) {
+                accountForAction = _sender;
+            }
+        }
+
+        require(_contract.call.gas(_gas)(_data));
+    }
+
+    function _defrayAndPassMessage(address _sender, address _contract, bytes _data, uint256 _gas) internal {
+        uint256 fee = (PASS_MESSAGE_GAS + _gas) * tx.gasprice;
```

Comment:
please use SafeMath here and in other places where arithmetic operations with variables are used.

Reply:
Thanks! Fixed